[UIK-3325] Update focus styles

j-mnizhek opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

✨ We’re updating the focus appearance to solve all the possible contrast issues with the background and to make it more modern and visually pleasant through animation ✨

**Note: Inputs, BulkTextarea, Textarea has no outline-offset, it's intended for now.**

Here’s what was changed:
1. Updated the common focus styles in the FlexBox component.
2. In the components where focus was previously styled locally using `box-shadow`, styles were rewritten to use `outline` and animation was added.
3. In components from the **FeatureHighlight** package, focus styles were slightly updated and animation was added to the `::before` pseudo-element.
4. The values of the `keyboard-focus-outline` variables were updated — transparency was completely removed.
5. Fixed tiny visual issue for checked Radio. When Radio has focused and checked state at the same time, border-color was dark. Now it has the color of the background.
6. Fixed tiny visual issue with border-radius in SPattern for invalid states.
7. Added a story for testing focus styles of inputs, Textarea and BulkTextarea, because their styles are set locally for now.
8. Added fallbacks for webkit browsers in FeatureHighlight components.

**To be done:**
1. ✅ To be done: invert focus styles for NoticeGlobal and NotoceBubble components. [Comment below with screenshots](https://github.com/semrush/intergalactic/pull/2338#issuecomment-3096039304)

**Proposal:**
✅ I think it will be a better decision for us to make some common focus styles for all components where they don’t come from the Box component. So that those components can reference the shared styles instead of defining them locally in a lot of places.

### Example

#### Current focus styles
<img width="247" height="98" alt="image" src="https://github.com/user-attachments/assets/c41cfb3a-8e84-4b3d-8c5c-7b911ac5d8be" />
<img width="543" height="135" alt="image" src="https://github.com/user-attachments/assets/3d1161ef-a38e-4371-9181-ee83494347eb" />
<img width="251" height="91" alt="image" src="https://github.com/user-attachments/assets/b236e067-6a61-4ee5-b416-98a484bec132" />

#### New focus styles
<img width="282" height="127" alt="image" src="https://github.com/user-attachments/assets/2fedc059-e5b8-44be-898c-ef57a42ce0f3" />
<img width="533" height="126" alt="image" src="https://github.com/user-attachments/assets/8f23ae1c-6d0d-4f75-b461-5a0aac07bcd2" />
<img width="332" height="61" alt="image" src="https://github.com/user-attachments/assets/884c40d0-c544-4258-8591-fddf031d99c7" />

### Video
One of the examples before:
https://github.com/user-attachments/assets/043e68f0-9e54-4610-91eb-3ddb7cd45187

One of the examples after:
https://github.com/user-attachments/assets/96ccfe4b-4106-4db0-8eec-bffc99c61503

Updated focus styles for **FeatureHighlight** package 😍
https://github.com/user-attachments/assets/098c38c4-9f8b-4cab-a5ad-c6c2856c1e9a

## How has this been tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- For example: -->
<!--- I have added unit tests -->
<!--- I have added Voice Over tests -->
<!--- Code cannot be tested automatically so I have tested it only manually -->

Manually with existing stories and examples in our documentation.

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Nice improve.

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have updated the documentation accordingly.
- [x] I have added changelog note to corresponding `CHANGELOG.md` file with planned publish date.
- [ ] I have added new tests on added of fixed functionality.

j-mnizhek commented on semcore/carousel/src/style/carousel.shadow.css:
One issue I couldn’t solve: in the carousel, the item’s focus outline always gets clipped at the top. However, if you open the item in a modal, the outline won’t be clipped

<img width="701" height="574" alt="Screenshot 2025-07-18 at 15 25 31" src="https://github.com/user-attachments/assets/4142f50b-016b-4b4f-826e-42f40d26f67e" />
<img width="1055" height="787" alt="Screenshot 2025-07-18 at 15 15 59" src="https://github.com/user-attachments/assets/58d2a17b-77d9-4f12-a56c-40b3679c07a8" />

ilyabrower commented on semcore/dropdown/src/style/dropdown.shadow.css:
```diff
@@ -67,13 +69,16 @@ SDropdownPopper {
   SDropdownItem[highlighted][isMenuItemCheckbox]:not(button) {
-    box-shadow: var(--intergalactic-keyboard-focus, 0px 0px 0px 3px rgba(0, 143, 248, 0.5)) inset;
+    outline: 2px solid var(--intergalactic-keyboard-focus-outline, #008ff8);
+    outline-offset: -2px;
```
not sure about -2

ilyabrower commented on semcore/dropdown/src/style/dropdown.shadow.css:
```diff
@@ -89,9 +94,11 @@ SDropdownPopper {
     &[isMenuItemCheckbox] {
-      padding-left: calc(var(--intergalactic-spacing-3x, 12px) * 2 + 16px + 2px); /* padding*2 (parent item and from checkbox to content) + checkboxWidth + border*2 */
+      padding-left: calc(var(--intergalactic-spacing-3x, 12px) * 2 + 16px + 2px);
+      /* padding*2 (parent item and from checkbox to content) + checkboxWidth + border*2 */
```
please turn off unnecessary formatting - half of your mr is unnecessary removing last line/adding indents or removing spaces.
here this comment should be on the same line or before

ilyabrower commented on semcore/feature-highlight/src/components/checkbox/checkbox.shadow.css:
```diff
@@ -1,38 +1,51 @@
+    mask: linear-gradient(0deg, #fff, #fff) padding-box, linear-gradient(0deg, #fff, #fff);
     mask-composite: exclude;
+    -webkit-mask: linear-gradient(#fff 0 0) padding-box, linear-gradient(#fff 0 0);
+    -webkit-mask-composite: xor;
```
why do you manually added webkit?

j-mnizhek replied:
Do you mean we have some automatic solution for it?

ilyabrower commented on semcore/inline-input/src/style/inline-input.shadow.css:
```diff
@@ -107,7 +107,10 @@ SAddon[disabled] {
 SUnderline:has(input:focus-visible) {
   border-bottom-color: var(--intergalactic-border-info-active, #006dca);
-  box-shadow: var(--intergalactic-keyboard-focus, 0px 0px 0px 3px rgba(0, 143, 248, 0.5));
+  outline: 2px solid var(--intergalactic-keyboard-focus-outline, #008ff8);
```
as we use outline, maybe we can use this styles from Box?

j-mnizhek replied:
It would be great ✨ I wrote a comment about this above https://github.com/semrush/intergalactic/pull/2338#issuecomment-3090448748

Or do you mean in this particular place?

j-mnizhek commented on semcore/dropdown-menu/src/style/dropdown-menu.shadow.css:
```diff
@@ -51,6 +54,7 @@ SDropdownMenuNesting[size='m'] {
   padding: 0;
 }
 
+/* What we need these rules for? */
```
It was my working comment. I cannot reproduce the case when these particular styles are applied to the nested menus. Plan to change them to outline styles ofc.

j-mnizhek commented on semcore/core/src/theme/light.json:
```diff
@@ -1534,44 +1534,44 @@
     },
     "keyboard-focus": {
-      "value": "0px 0px 0px 3px {keyboard-focus.outline}",
+      "value": "0px 0px 0px 3px rgba({keyboard-focus.outline}, 0.5)",
```
Yes, it's used by teams, we cannot remove it now without breaking something

sheila-semrush commented on semcore/base-components/src/components/flex-box/style/use-box.shadow.css:
```diff
@@ -1,10 +1,8 @@
+  outline: 2px solid var(--intergalactic-keyboard-focus-outline, #008ff8);
+  outline-offset: 2px;
   z-index: 1;
+  transition: all calc(var(--intergalactic-duration-extra-fast, 100) * 1ms) ease-in-out;
```
it seems it would be better to list specific attributes instead of "all", because now for example we have a side effect in table cells - fill has a transition on hover (and probably other controls that change styles on hover and other events)

j-mnizhek commented on semcore/input/src/style/input.shadow.css:
```diff
@@ -133,27 +136,32 @@ SInput[state='normal'] {
   &:has(SValue:focus-visible) SOutline {
     border-color: var(--intergalactic-border-info-active, #006dca);
-    box-shadow: var(--intergalactic-keyboard-focus, 0px 0px 0px 3px rgba(0, 143, 248, 0.5));
+    outline: 2px solid var(--intergalactic-keyboard-focus-outline, #008ff8);
```
Yeah, also was thinking about this option

sheila-semrush commented on semcore/feature-highlight/src/components/pills/pills.shadow.css:
```diff
@@ -33,25 +14,61 @@ SHighlightedPill:focus-visible {
-    margin: -4px;
-    padding: 4px;
+    margin: 0;
+    padding: 2px;
```
we'll have to change the paddings/sizes for FH slightly, right now the visual padding in FH components is bigger than in the default ones

<img width="418" height="178" alt="imagen" src="https://github.com/user-attachments/assets/264f0ade-a639-4425-b07a-2fc7c9cef151" />
<img width="377" height="153" alt="imagen" src="https://github.com/user-attachments/assets/5c22a017-28e5-4daa-a843-2a3cf38e0b1f" />

ilyabrower commented on semcore/d3-chart/src/style/plot.shadow.css:
```diff
@@ -3,8 +3,8 @@
-SPlot:active,
-SPlot:focus {
+
+SPlot:active {
```
not sure, we should remove focus here

ilyabrower replied:
this is a box, or, it should be ok, resolved

ilyabrower commented on semcore/date-picker/src/style/calendar.shadow.css:
```diff
@@ -70,7 +71,12 @@ SCalendarUnit {
+    outline-offset: 2px;
+    transition: outline-color calc(var(--intergalactic-duration-extra-fast, 100) * 1ms) ease-in-out,
+      outline-width calc(var(--intergalactic-duration-extra-fast, 100) * 1ms) ease-in-out,
+      outline-offset calc(var(--intergalactic-duration-extra-fast, 100) * 1ms) ease-in-out;
```
not sure we need this, it will come from Box

ilyabrower commented on semcore/feature-highlight/src/components/checkbox/checkbox.shadow.css:
```diff
@@ -1,38 +1,51 @@
 
-    background: var(--intergalactic-border-feature-highlight-active, linear-gradient(90deg, #ab6cfe, #008ff8)) padding-box, var(--intergalactic-keyboard-focus-feature-highlight-outline, linear-gradient(90deg, rgba(171, 108, 254, 0.5), rgba(0, 143, 248, 0.5))) border-box;
+    -webkit-mask: linear-gradient(#fff 0 0) content-box, linear-gradient(#fff 0 0);
+    -webkit-mask-composite: xor;
```
not sure, you need this -webkit properties because of postcss (or we should change some settings on it)

ilyabrower commented on semcore/feature-highlight/src/inner-components/button-trigger/buttonTrigger.shadow.css:
```diff
@@ -2,15 +2,39 @@ SHighlightedButtonTrigger:not([state="invalid"]):not([state="valid"]) {
   background-origin: border-box;
   background-clip: padding-box, border-box;
+  position: relative;
+  z-index: 1;
```
not sure about that

j-mnizhek replied:
We can remove this ofc if it won't affect the final look. I've already forgotten why I added this. But since I was trying different approaches for components in the FeatureHighlight package, it's a chance that these particular styles are not necessary, yep 👍 Need to check

ilyabrower commented on semcore/inline-input/src/style/inline-input.shadow.css:
```diff
@@ -116,7 +122,14 @@ SInlineInput[state='valid'] SUnderline {
 SInlineInput[state='valid'] SUnderline:has(input:focus-visible) {
   border-bottom-color: var(--intergalactic-border-success-active, #007c65);
-  box-shadow: var(--intergalactic-keyboard-focus-valid, 0px 0px 0px 3px rgba(0, 159, 129, 0.5));
+  z-index: -1;
```
do we really need this?

ilyabrower commented on semcore/inline-input/src/style/inline-input.shadow.css:
```diff
@@ -125,17 +138,23 @@ SInlineInput[state='invalid'] SUnderline {
 SInlineInput[state='invalid'] SUnderline:has(input:focus-visible) {
   border-bottom-color: var(--intergalactic-border-critical-active, #d1002f);
-  box-shadow: var(--intergalactic-keyboard-focus-invalid, 0px 0px 0px 3px rgba(255, 73, 83, 0.5));
+  z-index: -1;
```
and this

ilyabrower commented on stories/patterns/core/tests/focus-styles-in-all-inputs.tsx:
```diff
@@ -0,0 +1,79 @@
+import { Text } from '@semcore/typography';
+import React from 'react';
+
+const FocusStylesInAllInputs = () => {
```
this in not "all" inputs =)

j-mnizhek replied:
Definitely! I've added not all of them (but planned to), just some of them to compare focus styles.

j-mnizhek commented:
One more strange thing I've noticed: something strange is happening with the focus animation in InputFH (when you are leaving input with keyboard, looks like outline becoming smaller and flickers)
https://github.com/user-attachments/assets/fe24536c-8b00-4de4-a92e-7498f812d854

j-mnizhek commented:
> * focus offset in intearctive tags in InputTags can be removed (so that it's not cropped)

Applied new `focus-outline-mixin-no-offset` mixin here for tags inside InputTags

> * we might want to disable transition in components like DataTable cells or Menu items because it doesn't look right when navigating quickly with arrows
> * I think in DatePicker the day background also has an unwanted transition when selected (didn't look closely)

Agree. Removed transitions for CalendarUnit and menu items

> * we also might want to do something with transition in components where focus doesn't have an offset, because in my opinion it looks weird (incomplete)

Added mixin for focus without offset — `focus-outline-mixin-no-offset`. Applied it to Input, Textarea, BulkTextarea

> * Select: you don't see the left border indicator of a selected item when it's also highlighted

I'd left it as it is now, it's not a critical issue for us now
